### PR TITLE
first implementation of OAIPMHReader

### PR DESCRIPTION
**Pull request by wgresshoff**
:heart: Thank you for your contribution!

### Description

Add a new reader for harvesting of OAI-PMH based vocabularies like those at GND.

### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [x] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [x] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [x] I've added relevant test cases.
- [ ] I've added relevant documentation.
- [ ] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/howtos/i18n/).
- [ ] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [ ] I've NOT included third-party code (copy/pasted source code or new dependencies).
    * If you have added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect](https://github.com/orgs/inveniosoftware/teams/architects/members).

**Frontend**

- [ ] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines.
- [ ] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines.
- [ ] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines.

**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

### CODE REVIEW

File: invenio_vocabularies/datastreams/models.py
Comment (slint) [inline invenio_vocabularies/datastreams/models.py:1]: minor/nit: I would "inline" this file in the OAI-PMH reader, since it's tightly bound and probably the only place where it's going to be used.

I know the file is growing somewhat large, and the way to properly clean up would be to split more complex reader types into separate files while keeping all their logic/API classes together.

File: setup.cfg
Comment (slint) [inline setup.cfg:34]: minor/shelve: we can move the fork under `inveniosoftware` as well once we have a fully working datastream example :)

File: invenio_vocabularies/datastreams/readers.py
Comment (slint) [inline invenio_vocabularies/datastreams/readers.py:267]: minor/question: is `response.header` an XML tree Python object, or a custom internal class of Scythe?

It might be an option to also just return `{"response": response}`, and leave it to the transformer (or the next reader) to extract whatever information is needed.

**Review by slint — APPROVED** [review]
Looks good and very much in a working state. Some minor comments on organization and how to handle the new dependency.

File: invenio_vocabularies/datastreams/models.py
Comment (wgresshoff) [reply invenio_vocabularies/datastreams/models.py:11]: Is it possible to use the same package name? I changed the name to have a different package in PyPi.

File: tests/datastreams/test_datastreams.py
Comment (wgresshoff) [reply tests/datastreams/test_datastreams.py:156]: I was aware of the danger but wanted a quick solution. I have to investigate the possibilities. Possibly using pytest-httpserver may help here.

File: setup.cfg
Comment (wgresshoff) [reply setup.cfg:34]: Sure, if we don't need a Pypi package we should.

File: invenio_vocabularies/datastreams/readers.py
Comment (wgresshoff) [reply invenio_vocabularies/datastreams/readers.py:267]: Yes, `response.header` is an internal class (looked simply like a string at first look), as is `response`, too. If that's no problem you're right. I'll change that.

File: invenio_vocabularies/datastreams/models.py
Comment (wgresshoff) [reply invenio_vocabularies/datastreams/models.py]: This is changed.

File: invenio_vocabularies/datastreams/models.py
Comment (wgresshoff) [reply invenio_vocabularies/datastreams/models.py:1]: Done. Regarding the split into several files: I think so, too. We handle refactorings like this in different issues and PR's, what do you think?

File: tests/datastreams/test_datastreams.py
Comment (wgresshoff) [reply tests/datastreams/test_datastreams.py]: I replaced the code with a mocking solution based on `pytest_httpserver`.

File: invenio_vocabularies/datastreams/readers.py
Comment (ptamarit) [inline invenio_vocabularies/datastreams/readers.py:274]:
```suggestion
            raise NotImplementedError(
                "OAIPMHReader does not support being chained after another reader"
            )
```

This is what I did in the two readers I recently implemented, so that if somebody mistakenly defines a reader before this one, they will get a clear exception instead of silently doing nothing.

File: setup.cfg
Comment (ptamarit) [inline setup.cfg:34]: Do we want to release `invenio-oaipmh-scythe` to PyPI before merging or are we happy to keep a Git dependency?

File: setup.cfg
Comment (ptamarit) [inline setup.cfg:41]: @slint had mentioned a few libraries in the [`RORHTTPReader` Pull Request](https://github.com/inveniosoftware/invenio-vocabularies/pull/315#discussion_r1608132835).
I guess the decision on which library to use does not matter too much (since it's test-only), but I was curious about how you chose this specific one.

File: tests/datastreams/test_datastreams.py
Comment (ptamarit) [inline tests/datastreams/test_datastreams.py:251]: Since you are testing `OAIPMHReader` specifically, I would say that defining a full config/factory here is not necessary.
You could:
1. Move these tests from `tests/datastreams/test_datastreams.py` to `tests/datastreams/test_readers.py`
2. Change the test to look more like other reader tests (calling directly the constructor and then the `read` method).
3. Remove the definition of the OAIPMHReader in `tests/datastreams/conftest.py`

File: tests/datastreams/test_datastreams.py
Comment (ptamarit) [inline tests/datastreams/test_datastreams.py:260]: I'm not understanding why we use a `for` loop is it's to directly `break` out of it.

File: tests/datastreams/test_datastreams.py
Comment (ptamarit) [inline tests/datastreams/test_datastreams.py:299]:
```suggestion
    with pytest.raises(ReaderError):
        next(stream_iter)
```

This is what I used in a few tests called `test_ror_http_reader_...`, I think it should work.

**Review by ptamarit — APPROVED** [review]
Looks good to me.
I have a few minor comments, and I'm also _wondering if all the commits should be squashed into one_.

File: setup.cfg
Comment (wgresshoff) [reply setup.cfg:34]: I thought so, too, but it's just a temporary solution until we completely get rid of Python 3.9.

File: setup.cfg
Comment (wgresshoff) [reply setup.cfg:41]: I did not look too deep for a library. `Response` (in my opinion) is heavily written for testings based on `requests`, while `oaipmh_scythe` is based on `httpx`. It seemed a good idea to use a library that directly hooks into the protocol, thus working independently from any library used to implement http calls.

File: tests/datastreams/test_datastreams.py
Comment (wgresshoff) [reply tests/datastreams/test_datastreams.py]: Done.

**Pull request merged**
